### DISCUSSION TITLE
Experiencing Lag/Crash on Emoji Picker Only IOS

### DISCUSSION THREAD

**Discussion by lcundiff** [hm21/pro_image_editor#259] [Q&A]
Since emoji lag isn't listed as an issue on the repo, I figured something with our configs is off, and wanted to see if anyone else has experienced a similar issue with the emoji picker lagging tremendously when opening emoji editor only on IOS? 

What happens is that the emoji picker lags on opening and when its added will lag anytime an update occurs - which could be adding another emoji, shape, or even clicking an editor tool. So I think its lagging from the image being generated in background with the emoji and for some reason it doesn't like a combination of our settings and the emoji on IOS. 

Been debugging for a couple of weeks, but thought I'd start a discussion in case others have experienced it and to help future devs with same issue potentially. 

```
                      configs: ProImageEditorConfigs(
                          imageGenerationConfigs: ImageGenerationConfigs(
                          processorConfigs: const ProcessorConfigs(
                            processorMode: ProcessorMode.auto,
                            maxConcurrency: 2,
                            numberOfBackgroundProcessors: 2,
                          ),
                          generateInsideSeparateThread: true,
                          captureOnlyDrawingBounds: false,
                          captureOnlyBackgroundImageArea: true, 
                          allowEmptyEditCompletion: false,
                          singleFrame: true,
                          outputFormat: OutputFormat.png,
                          pngLevel: 1,
                        ),
```
Also, I am not sure what is recommended for max Concurrency or numberOfBackgroundProcessors, but if there are any tips on choosing these numbers as well, that would be helpful! 

One last note, our app also lags and freezes when drawing a thick line with the draw tool, but I tested this on the web demo, and the thick line only lags a little bit. So once again, I think our configuration settings may be off. So I will update with any discoveries with different combinations.

**Comment by hm21** [comment]
@lcundiff, your configurations don’t appear problematic for that issue to me. However, I typically recommend avoiding setting any values in `processorConfigs` unless there is a specific need for a few devices. In most cases, you may not know the exact devices your users will be using, so it’s best to rely on default settings. The default configuration is `maxConcurrency: 1`, and `numberOfBackgroundProcessors` is determined by the editor based on the device's processor capabilities. You can review this calculation in the [processor_helper](https://github.com/hm21/pro_image_editor/blob/stable/lib/utils/content_recorder.dart/utils/processor_helper.dart) class.

On mobile devices, setting `maxConcurrency` higher than 1 may overload the CPU, especially if there are insufficient free CPU cores. This can significantly slow down the generation process and negatively impact performance.

The lag issue with the emoji picker appears to stem primarily from Flutter's graphics engine, `Impeller`. I have discussed this in some detail in [this comment](https://github.com/hm21/pro_image_editor/issues/122#issuecomment-2182412179). The lag is especially noticeable when using a custom font—such as `GoogleFonts.noToColorEmoji`—instead of the system default font. Additionally, there is an open issue in the `emoji_picker_flutter` package, which you can see [here](https://github.com/Fintasys/emoji_picker_flutter/issues/205).
